feat: Create scripts to help us format Airbyte generated dbt

jitoquinto opened the pull request:
Alright y'all, here's a chonky PR which adds a couple new scripts.

- `utils/dbt_format_utils.py`: This is a python script that does most of the heavy lifting. It can:
  -  rename base tables to add the  '_base` suffix to all files in a folder
  - create matview sql files from a directory of base tables
  - download a cacher script to local
    - This is used to download the script to pull down dbt from Airbyte, so now we can do everything through the helper script
  - restructure airbyte default normalization to cta structure (this moves files into correct directory structure + modifies dbt files to be more of what we expect)
- `cta_dbt_helper.sh`: this is a shell script that kinda just works as a wrapper for the python script. it uses gum to prompt for user input and then does the things. I figured this was an easier way to interact rather than having to read up doc and figure out all the right things to type when calling python

I learned that modifying dbt files is a pain with both sed and python so hopefully airbyte doesnt change up their normalization and breaks these 🙃 

Would love if yall could also take the `cta_dbt_helper.sh` helper script for a run and give feedback on what can be improved and if there are things i can add to make it more clear what should be inputted!

TODO: 

- [x] Figure out why test generation script isnt working @jackwelty would love some support here if you have some time!
- [x] Modify README with some instructions on using the scripts
- [x] Tun black and hope it doesnt mess up the hardcoded sql strings 🤞

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,329 @@
+
+def main():
+    parser = argparse.ArgumentParser()
+    sub_parsers = parser.add_subparsers(
```
o shit subparsers, this is new to me 👀

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,329 @@
+        )
+
+
+# Function to grab Airbyte Generated dbt and move into CTA structured folder (✨Progify it✨)
```
🐸 🐸 🐸 🐸

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,329 @@
+
+            # Move files to target path
+            for file_path in files_to_move:
+                file_name = file_path.split("/")[-1]
```
I have recently learned through the magic of ChatGPT that a slightly better practice here is to use `os.path.basename()` instead if splitting the path string! I guess it's more platform independent, and it's unclear if that's actually a relevant concern, but bp's are nice 👀 

```suggestion
                file_name = os.path.basename(file_path)
```

In general I highly recommend asking ChatGPT for bp's regarding file operations - there's a lot you can do more cleanly with `os` (this is all news to me via my own heftychonk PR w the Cloud Function).

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,329 @@
+
+# Function to modify ctes and incremental models. For these its just gonna be
+# swapping out the config section, and changing source to cta.
+def modify_dbt_models(path_to_models, model_type):
```
general request: Can we add docstrings to all the functions? I know I am a broken record here but chatgpt does this VERY fast, and it will make the code more readable/maintanable/etcetc

jitoquinto replied:
yup can do 👍

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,329 @@
+        '    unique_key = "_airbyte_ab_id"\n'
+        ") }}\n"
+    )
+    stop_skip_pattern = "-- SQL model" if model_type == "cte" else "-- Final base SQL"
```
yeah lol so there will very possibly come a time when Airbyte changes something and this strat won't work anymore 🫠🫠🫠 Probably this strat makes sense for now, but wonder if A) there are any other ways we can logically achieve the same thing, and B) we should probably flag this in a comment or the readme or have some error handling or something (like, maybe raise an exception if neither of these patterns are present?)

jitoquinto replied:
yeahhh this definitely a problem. The other thing that came up while testing this is that the macros change... i dont really know what the best approach for fixing that other than running the dbt and seeing what fails. I honestly think the best solution is to make our own version of the default normalizer or i guess just live with the fact that this will just need constant upkeep. But yeah i can add some kind of a check here

emily-flambe replied:
> I honestly think the best solution is to make our own version of the default normalizer

This would be the dream! Fisher and I derped around once in the Airbyte repo to try and find the code they use to run default normalizations. We were more or less unsuccessful... I think whatever process they use is buried very deep within the codebase, and maybe even in like Java or something. I wonder how much of an undertaking this would be? It would make our processes much more resilient to use Airbyte ONLY up until the point where it delivers raw JSON, and then we could have code that flattens the JSON and creates the dbt models we crave... It seems doable, at least in theory 🧐 This would be a heftychomnk project though so I'd like us to discuss with Fisher. I'll draft a ticket tho 😤

emily-flambe commented on cta_dbt_helper.sh:
```diff
@@ -0,0 +1,188 @@
+        # Get Cacher creds from user
+        CACHER_API_KEY=$(gum input --placeholder "Enter your Cacher API Key -> https://app.cacher.io/enter?action=view_api_creds")
+        CACHER_API_TOKEN=$(gum input --placeholder "Enter your Cacher API Token -> https://app.cacher.io/enter?action=view_api_creds")
+        CACHER_SNIPPET_GUID="2b77280d537736f980f9"
```
Let's move this to a Github Secret 👀

emily-flambe replied:
Hmm, yeah you are right I kinda forgot that's how Github Secrets work, and you're also right this isn't actually a Secret-Secret - can we just add some commentary explaining what the script does so if a curious onlooker sees this they will know what we are doing?

emily-flambe reviewed (commented):
Haven't taken this for a test drive yet but left some comments for you to chew on @jitoquinto 👀 👀 👀 👀  lots of exquisite stuff happening here, so excited for this to be non-scuffed lol

emily-flambe commented:
@jitoquinto when selecting the Airbyte environment, should it appear like this, with `-e dev` and `prod` as the options? Or should it just be `dev` and `prod`? Maybe it doesn't matter? 
<img width="311" alt="image" src="https://user-images.githubusercontent.com/47834372/233850461-2dd87648-63cd-4d41-831b-b1f0085c5adc.png">

emily-flambe commented:
@jitoquinto I tried to run this to grab the dbt from workspace 801 in dev and hit this error:

```
emilycogsdill@MacBook-Pro-3 dbt-cta % sh cta_dbt_helper.sh
cta_dbt_helper.sh: line 19: cd: /Users/emilycogsdill/Documents/GitHub/dbt-cta/.cta: No such file or directory
Cacher Snippet Copy Airbyte-Generated dbt from VM to GCS (via IAP Tunnel) downloaded to copy_airbyte_workspace.sh
Starting job to copy Airbyte workspace..
cta_dbt_helper.sh: line 36: ./.cta/copy_airbyte_workspace.sh: No such file or directory
Failed to grab normalization dbt for workspace: 801 from -e dev Airbyte
```

rip

emily-flambe commented:
@jitoquinto (sorry I know this is a lot of Github to wake up to on a Monday lol!)

I was working on Empower Phase 2 and decided to use ChatGPT to simplify the process of modifying the full refresh models - not sure how much this overlaps with what you have, but please crib anything that is useful to you! [modify_full_refresh_models.py](https://gist.github.com/emily-flambe/caae5b5e873946aab6399bc3d560768e#file-modify_full_refresh_models-py)

Note also it does some cleanup to remove a config called "post_hook" that is visually oppressive (generated for the empower sync - look in those 1_cta_full_refresh files if you wanna see what I mean). I am pretty sure we do not need it 🤔

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,329 @@
+            with open(model_file_path, "r") as model_file:
+                for line in model_file:
+                    # Dont start writing to file if were still in config section
+                    if "-- Final base SQL" in line:
```
OK yeah lol so the code I slapped together yesterday is mostly duplicative / not quite as elegant as yours - the one thing I would suggest is adjusting the logic for separating all the config stuff at the top from the SQL model itself. I think a strat that might be more resilient than looking for `-- Final base SQL` would be to treat the config section as a "block" that starts with `{{ config(` and ends with `) }}` - [here's how I twerked that logic in my code.](https://gist.github.com/emily-flambe/caae5b5e873946aab6399bc3d560768e#file-modify_full_refresh_models-py-L60-L64)

jitoquinto replied:
ooh yeah thats a good idea, Ill switch it to use those strings

emily-flambe commented:
I can at least confirm that the function for creating matviews from base tables works a charm!!! I pulled it out into a standalone script ([gist](https://gist.github.com/emily-flambe/83eb98c6879392ef7aa975c0484231d3)) to take that individual piece for a test drive. 🎉

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,329 @@
+        print(f"- name: {table_name}_base")
+
+
+def create_matview_dbt_files_from_base(base_tables_path, output_path):
```
I can't quickly pinpoint where in the code this should be revised, but I think it will be easier for the matviews to query from {{ ref('TABLE_base') }} instead of {{ source('cta','TABLE_base') }} - the latter strat requires adding the base tables to sources.yml, and we can just skip that step by using ref instead 👀

jitoquinto commented:
> @jitoquinto when selecting the Airbyte environment, should it appear like this, with `-e dev` and `prod` as the options? Or should it just be `dev` and `prod`? Maybe it doesn't matter? <img alt="image" width="311" src="https://user-images.githubusercontent.com/47834372/233850461-2dd87648-63cd-4d41-831b-b1f0085c5adc.png">

🤔 weird im not seeing it when i run it on my CLI. I wonder whats different? this is what is being ran the `-e` is a flag so that echo interprets the `\n` as new line in the `ALLOWED_ENVIRONMENTS` var
```shell
ENVIRONMENT=$(echo -e $ALLOWED_ENVIRONMENTS | gum filter --placeholder "Which Airbyte Environment should be used?")
```

jitoquinto commented:
> @jitoquinto I tried to run this to grab the dbt from workspace 801 in dev and hit this error:
> 
> ```
> emilycogsdill@MacBook-Pro-3 dbt-cta % sh cta_dbt_helper.sh
> cta_dbt_helper.sh: line 19: cd: /Users/emilycogsdill/Documents/GitHub/dbt-cta/.cta: No such file or directory
> Cacher Snippet Copy Airbyte-Generated dbt from VM to GCS (via IAP Tunnel) downloaded to copy_airbyte_workspace.sh
> Starting job to copy Airbyte workspace..
> cta_dbt_helper.sh: line 36: ./.cta/copy_airbyte_workspace.sh: No such file or directory
> Failed to grab normalization dbt for workspace: 801 from -e dev Airbyte
> ```
> 
> rip

rip indeed, where did the `copy_airbyte_workspace.sh` script get downloaded to? I recently moved this script back into the utils folder and im trying to figure out what is happening with the `ROOT_PATH` var now 🙃

emily-flambe commented:
> @emily-flambe rip indeed, where did the `copy_airbyte_workspace.sh` script get downloaded to? I recently moved this script back into the utils folder and im trying to figure out what is happening with the `ROOT_PATH` var now 🙃

🤔  I actually completely forgot about that script, I'm pretty sure I haven't moved it anywhere in recent memory 🤔 🤔 🤔 it's not something like [this](https://snippets.cacher.io/snippet/be07ef075f7db860d2b7) is it?

emily-flambe commented:
@jitoquinto I know you're ready to be done with this but I do have one small feature request: can we ensure all dbt model names are all lowercase? The dbt test generation script puts all the test names in lowercase, and the case needs to match the filenames, so it seems easiest to just force everything to lowercase here - even some kind of blunt strategy e.g. iterating through all files in the model directory and os.renaming them to a lowercased version of their filenames. Could you possibly add that in? 🙏 🙏 🙏  or if you're REALLY done with this we can spin that off into a 1-point tech debt ticket lol, it's not really a broken thing, just a thing that's easy to mess up 🧐

jitoquinto commented:
> @jitoquinto I know you're ready to be done with this but I do have one small feature request: can we ensure all dbt model names are all lowercase? The dbt test generation script puts all the test names in lowercase, and the case needs to match the filenames, so it seems easiest to just force everything to lowercase here - even some kind of blunt strategy e.g. iterating through all files in the model directory and os.renaming them to a lowercased version of their filenames. Could you possibly add that in? 🙏 🙏 🙏 or if you're REALLY done with this we can spin that off into a 1-point tech debt ticket lol, it's not really a broken thing, just a thing that's easy to mess up 🧐

yeah sure i can add that. qq is there a case where they arent lowercase?? at least from from what i can remember all airbyte generated dbt has had lowercase file names

jitoquinto commented:
> > @jitoquinto I know you're ready to be done with this but I do have one small feature request: can we ensure all dbt model names are all lowercase? The dbt test generation script puts all the test names in lowercase, and the case needs to match the filenames, so it seems easiest to just force everything to lowercase here - even some kind of blunt strategy e.g. iterating through all files in the model directory and os.renaming them to a lowercased version of their filenames. Could you possibly add that in? 🙏 🙏 🙏 or if you're REALLY done with this we can spin that off into a 1-point tech debt ticket lol, it's not really a broken thing, just a thing that's easy to mess up 🧐
> 
> Oh actually this would also require making sure all the references within all the models are _also_ lowercase, so maybe annoying enough to get its own tech debt ticket - what do you think? Super-easy, or make this into a ticket?

yeah lets just do another ticket, i think that would push it to me continuing to work on this tomorrow and id like to be done with it today!

emily-flambe commented:
Cool, I'll draft that tomorrow! Can salute emoji

On Mon, Apr 24, 2023, 3:03 PM Jesus Toquinto ***@***.***>
wrote:

> @jitoquinto <https://github.com/jitoquinto> I know you're ready to be
> done with this but I do have one small feature request: can we ensure all
> dbt model names are all lowercase? The dbt test generation script puts all
> the test names in lowercase, and the case needs to match the filenames, so
> it seems easiest to just force everything to lowercase here - even some
> kind of blunt strategy e.g. iterating through all files in the model
> directory and os.renaming them to a lowercased version of their filenames.
> Could you possibly add that in? 🙏 🙏 🙏 or if you're REALLY done with
> this we can spin that off into a 1-point tech debt ticket lol, it's not
> really a broken thing, just a thing that's easy to mess up 🧐
>
> Oh actually this would also require making sure all the references within
> all the models are *also* lowercase, so maybe annoying enough to get its
> own tech debt ticket - what do you think? Super-easy, or make this into a
> ticket?
>
> yeah lets just do another ticket, i think that would push it to me
> continuing to work on this tomorrow and id like to be done with it today!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/community-tech-alliance/dbt-cta/pull/124#issuecomment-1520822212>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALM6KBDQBSKW2R55OHKPGRTXC3TA7ANCNFSM6AAAAAAXHENWLY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

emily-flambe commented on utils/dbt_format_utils.py:
```diff
@@ -0,0 +1,365 @@
+                        write_to_file = False
+                        outfile.write(f"from {{{{ source('cta','{table_name}') }}}}")
+                    if write_to_file:
+                        if not "_airbyte" in line:
```
@jitoquinto I think we want to keep the `_airbyte_*_hashid` fields in there since those are supposed to be the unique primary keys, at least by default until we go in and figure out the actual UPKs for each table - can we modify this logic so it yeets all the _airbyte fields _except_ the hashid?

emily-flambe replied:
Actually I wonder if the easiest strat is to just do something like

```
select
    *
EXCEPT _airbyte_ab_id, _airbyte_emitted_at, _airbyte_normalized_at
from .....
```

🤔 whatever you think is easiest here is fine with me, but we do want them hashids 👀

jitoquinto replied:
@emily-flambe we want this hash_id even for matviews?

jitoquinto commented:
@emily-flambe okay i think i've addressed most of these if you would like to re-review/take the script for another go??

emily-flambe commented:
Hmm I am still hitting this error when trying to download the Airbyte workspace: 

``` 
(dbt-cta) emilycogsdill@Elsas-MBP-553 dbt-cta % sh ./utils/cta_dbt_helper.sh

./utils/cta_dbt_helper.sh: line 23: cd: /Users/emilycogsdill/Documents/GitHub/dbt-cta/.cta: No such file or directory
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project. You can set PIPENV_IGNORE_VIRTUALENVS=1 to force pipenv to ignore that environment and create its own instead. You can set PIPENV_VERBOSITY=-1 to suppress this warning.
Cacher Snippet Copy Airbyte-Generated dbt from VM to GCS (via IAP Tunnel) downloaded to copy_airbyte_workspace.sh
Starting job to copy Airbyte workspace..
./utils/cta_dbt_helper.sh: line 40: ./.cta/copy_airbyte_workspace.sh: No such file or directory
Failed to grab normalization dbt for workspace: 801 from -e dev Airbyte
```

What is the significance of `.cta` and why does the script cd into it? I am successfully downloading the cacher script - it's ending up in the top level of dbt-cta, and then it looks like the dbt helper is looking for it at `./.cta/copy_airbyte_workspace.sh`, which is why it's failing to locate it 🧐

jitoquinto commented:
@emily-flambe ahh okay! i forgot that directory doesnt exist yet so gotta create it 😓. pushed a fix. I made that directory to hold artifacts that we dont want to push to github. So this _should_ create that directory and download the script there. And i added the `.cta` directory to gitignore so any files in there dont get pushed to gh

kanelouise reviewed (changes requested):
I left a few comments for review, but overall this is a pleasant upgrade from our former shell script. Great job, can't wait to use it!

jitoquinto commented on utils/README.md:
````diff
@@ -10,44 +12,60 @@ Specifically, it makes the dbt files:
 
 ```
-sh ../utils/init_dbt
+sh ./utils/cta_dbt_helper.sh
````
ahh good catch!

kanelouise commented on utils/cta_dbt_helper.sh:
```diff
@@ -0,0 +1,199 @@
+            exit 1
+        fi
+    done
+    gum confirm "Confirm the Base models path is correct: $ROOT_PATH/$OUTPUT_DIR_NAME" || exit 1
```
🍝

emily-flambe commented on utils/cta_dbt_helper.sh:
```diff
@@ -0,0 +1,199 @@
+
+    # Get target path for directory where matviews should be created
+    while [ -z "$OUTPUT_DIR_NAME" ]; do
+        OUTPUT_DIR_NAME=$(gum input --prompt "Enter the directory name where matviews should be placed: " --placeholder "(ex. dbt-cta/actblue/2_partner_matviews/)")
```
@jitoquinto can we revise the placeholder here so it includes the `models` subdirectory? 🙏 

```suggestion
        OUTPUT_DIR_NAME=$(gum input --prompt "Enter the directory name where matviews should be placed: " --placeholder "(ex. dbt-cta/actblue/models/2_partner_matviews/)")
```

jitoquinto commented:
@emily-flambe @kanelouise history is starting to get a bit confusing with this PR https://github.com/community-tech-alliance/dbt-cta/pull/132#pullrequestreview-1407728468 since its branched from this branch (seeing some files getting deleted) but also now both branches are out of sync with main. i think i've addressed all the comments, could we merge this in and i can start a new PR with any other requested changes? (ill sync this branch with main first so that @kanelouise can pull in changes to their branch and hopefully things will be okay)

emily-flambe commented:
> @emily-flambe @kanelouise history is starting to get a bit confusing with this PR [#132 (review)](https://github.com/community-tech-alliance/dbt-cta/pull/132#pullrequestreview-1407728468) since its branched from this branch (seeing some files getting deleted) but also now both branches are out of sync with main. i think i've addressed all the comments, could we merge this in and i can start a new PR with any other requested changes? (ill sync this branch with main first so that @kanelouise can pull in changes to their branch and hopefully things will be okay)

Yep I'm in favor of resolving any github weirdness, merging this in, and requesting new changes in a ticket if needed!

kanelouise commented:
FWIW, @jitoquinto and @emily-flambe the only files i added to [DE-240's PR ](https://github.com/community-tech-alliance/dbt-cta/pull/132)were the ones I actually meant to edit. I *think* that should prevent us from having a git mess.

jitoquinto commented:
Okay cool! Can definitely help out if git does get confused
